wcs-functions.php

```diff
@@ -149,62 +154,36 @@ function wcs_create_subscription( $args = array() ) {
-		$subscription_data['post_parent'] = absint( $args['order_id'] );
-	}
-
-	if ( ! is_null( $args['customer_note'] ) && ! empty( $args['customer_note'] ) ) {
-		$subscription_data['post_excerpt'] = $args['customer_note'];
-	}
-
-	// Only set the status if creating a new subscription, use wcs_update_subscription to update the status
+	$subscription = new \WC_Subscription();
 	if ( $args['status'] ) {
-		if ( ! in_array( 'wc-' . $args['status'], array_keys( wcs_get_subscription_statuses() ) ) ) {
-			return new WP_Error( 'woocommerce_invalid_subscription_status', __( 'Invalid subscription status given.', 'woocommerce-subscriptions' ) );
-		}
-		$subscription_data['post_status']  = 'wc-' . $args['status'];
-	}
+		$subscription->set_status( $args['status'] );
+	}
+	$subscription->set_customer_note( $args['customer_note'] ?? '' );
+	$subscription->set_customer_id( $args['customer_id'] );
+	$subscription->set_date_created( $args['date_created'] );
+	$subscription->set_created_via( $args['created_via'] );
+	$subscription->set_currency( $args['currency'] );
+	$subscription->set_prices_include_tax( 'no' !== $args['prices_include_tax'] );
+	$subscription->set_billing_period( $args['billing_period'] );
+	$subscription->set_billing_interval( absint( $args['billing_interval'] ) );
+	$subscription->set_schedule_start( $args['start_date'] );
 
-	$subscription_id = wp_insert_post( apply_filters( 'woocommerce_new_subscription_data', $subscription_data, $args ), true );
-
-	if ( is_wp_error( $subscription_id ) ) {
-		return $subscription_id;
+	if ( $args['order_id'] > 0 ) {
+		$subscription->set_parent_id( $args['order_id'] );
 	}
 
-	// Default order meta data.
-	update_post_meta( $subscription_id, '_order_key', wcs_generate_order_key() );
-	update_post_meta( $subscription_id, '_order_currency', $args['currency'] );
-	update_post_meta( $subscription_id, '_prices_include_tax', $args['prices_include_tax'] );
-	update_post_meta( $subscription_id, '_created_via', sanitize_text_field( $args['created_via'] ) );
-
-	// add/update the billing
-	update_post_meta( $subscription_id, '_billing_period', $args['billing_period'] );
-	update_post_meta( $subscription_id, '_billing_interval', absint( $args['billing_interval'] ) );
-
-	update_post_meta( $subscription_id, '_customer_user', $args['customer_id'] );
-	update_post_meta( $subscription_id, '_order_version', $args['order_version'] );
-
-	update_post_meta( $subscription_id, '_schedule_start', $args['start_date'] );
+	$subscription->save();
 
 	/**
 	 * Filter the newly created subscription object.
 	 *
 	 * @since 2.2.22
 	 * @param WC_Subscription $subscription
 	 */
-	$subscription = apply_filters( 'wcs_created_subscription', wcs_get_subscription( $subscription_id ) );
+	$subscription = apply_filters( 'wcs_created_subscription', $subscription );
+
+	// we need to fetch the subscription from the database as the current object state doesn't match the loaded state.
+	$subscription = wcs_get_subscription( $subscription );
```

Comment:
This `wcs_get_subscription( $subscription );` will need to return to its original place in the `apply_filters` above.


```
$subscription = apply_filters( 'wcs_created_subscription', wcs_get_subscription( $subscription_id ) );
```

This is because if a 3rd party hooks onto the filter and alters the subscription in some way, any changes they made would be overridden by the `wcs_get_subscription()`.